lib/util/QueuePool.js

```diff
@@ -0,0 +1,97 @@
+"use strict";
+let Promise = require("bluebird");
+let Queue = require("./Queue");
+
+// A Queue Pool is a queue which is backed by a pool of queues which can be serviced
+// concurrently. The number of items which can be processed concurrently is the size
+// of the queue. The QueuePool always operates in a FIFO manner, even when all queues
+// are occupied.
+class QueuePool {
+
+    // Construct a new Queue Pool.
+    // This consists of multiple queues. Items will be inserted into
+    // the first available free queue. If no queue is free, items will
+    // be put in a FIFO overflow queue. You can also use an index when
+    // enqueuing to override this behaviour.
+    constructor(poolSize, fn) {
+        if (poolSize < 1) {
+            throw new Error("Pool size must be at least 1");
+        }
+        this.size = poolSize;
+        this.fn = fn;
+        this.queues = [];
+        for (let i = 0; i < poolSize; i++) {
+            this.queues.push(new Queue(fn));
+        }
+        this.overflow = new Queue(this._overflow.bind(this));
+    }
+
+    // Add an item to the queue. ID and item are passed directly to the Queue.
+    // Index is optional and should be between 0 ~ poolSize-1. It determines
+    // which queue to put the item into.
```

Comment:
I would mention that specifying an index also ignores the overflow, and enqueues an item immediately